File: src/lib/sets/destinySetConstraints.test.ts
import { describe, expect, it } from "vitest";

import {
  assertSetCompositionAllowed,
  assertSetExoticExclusivity,
  assertSetItemAllowed,
  kindFromEntityStore,
  setAlreadyHasExotic,
  setItemMetaFromCatalog,
  setItemMetaFromManifestCategory,
} from "./destinySetConstraints";

describe("assertSetItemAllowed", () => {
  it("rejects invalid slots for set type", () => {
    const r = assertSetItemAllowed("weapon", "helmet", {
      kind: "weapon",
      equipmentSlot: "Kinetic",
    });
    expect(r.ok).toBe(false);
    if (!r.ok) expect(r.reasons[0]).toMatch(/not valid/i);
  });

  it("rejects armor in a weapon set", () => {
    const r = assertSetItemAllowed("weapon", "primary", {
      kind: "armor",
      equipmentSlot: "Helmet",
    });
    expect(r.ok).toBe(false);
    if (!r.ok) expect(r.reasons.join(" ")).toMatch(/weapons/i);
  });

  it("allows unknown kind on armor set (legendary armor store miss)", () => {
    expect(
      assertSetItemAllowed("armor", "helmet", { kind: "unknown" }),
    ).toEqual({ ok: true });
  });

  it("rejects Kinetic weapon in special (Energy) slot", () => {
    const r = assertSetItemAllowed("weapon", "special", {
      kind: "weapon",
      equipmentSlot: "Kinetic",
    });
    expect(r.ok).toBe(false);
    if (!r.ok) expect(r.reasons.join(" ")).toMatch(/Energy/i);
  });

  it("allows matching Kinetic weapon in primary", () => {
    const r = assertSetItemAllowed("weapon", "primary", {
      kind: "weapon",
      equipmentSlot: "Kinetic",
      isExotic: false,
    });
    expect(r).toEqual({ ok: true });
  });

  it("allows exotic weapon in primary", () => {
    const r = assertSetItemAllowed("weapon", "primary", {
      kind: "exotic_weapon",
      equipmentSlot: "Kinetic",
      isExotic: true,
    });
    expect(r).toEqual({ ok: true });
  });

  it("rejects wrong armor piece", () => {
    const r = assertSetItemAllowed("armor", "helmet", {
      kind: "armor",
      equipmentSlot: "Legs",
    });
    expect(r.ok).toBe(false);
  });

  it("allows matching helmet", () => {
    expect(
      assertSetItemAllowed("armor", "helmet", {
        kind: "armor",
        equipmentSlot: "Helmet",
      }),
    ).toEqual({ ok: true });
  });

  it("rejects legendary weapon in pair exotic_weapon", () => {
    const r = assertSetItemAllowed("pair", "exotic_weapon", {
      kind: "weapon",
      isExotic: false,
      equipmentSlot: "Kinetic",
    });
    expect(r.ok).toBe(false);
    if (!r.ok) expect(r.reasons.join(" ")).toMatch(/exotic weapon/i);
  });

  it("allows exotic weapon in pair exotic_weapon", () => {
    expect(
      assertSetItemAllowed("pair", "exotic_weapon", {
        kind: "exotic_weapon",
        isExotic: true,
      }),
    ).toEqual({ ok: true });
  });

  it("rejects non-exotic armor in pair exotic_armor", () => {
    const r = assertSetItemAllowed("pair", "exotic_armor", {
      kind: "armor",
      isExotic: false,
      equipmentSlot: "Helmet",
    });
    expect(r.ok).toBe(false);
  });

  it("rejects weapon hash meta in mod set", () => {
    const r = assertSetItemAllowed("mod", "mod:1", {
      kind: "weapon",
      equipmentSlot: "Kinetic",
    });
    expect(r.ok).toBe(false);
  });

  it("allows mods in mod set", () => {
    expect(
      assertSetItemAllowed("mod", "mod:42", { kind: "mod" }),
    ).toEqual({ ok: true });
  });

  it("allows fashion freely when slot valid", () => {
    expect(
      assertSetItemAllowed("fashion", "ghost", { kind: "unknown" }),
    ).toEqual({ ok: true });
  });

  it("skips bucket check when equipmentSlot unknown", () => {
    expect(
      assertSetItemAllowed("weapon", "primary", {
        kind: "weapon",
        equipmentSlot: null,
      }),
    ).toEqual({ ok: true });
  });
});

describe("assertSetExoticExclusivity", () => {
  it("blocks a second exotic weapon in a weapon set", () => {
    const r = assertSetExoticExclusivity({
      setType: "weapon",
      otherItems: [
        {
          slot: "primary",
          meta: {
            kind: "exotic_weapon",
            isExotic: true,
            name: "Witherhoard",
            equipmentSlot: "Kinetic",
          },
        },
      ],
      candidate: {
        slot: "heavy",
        meta: {
          kind: "exotic_weapon",
          isExotic: true,
          name: "Gjallarhorn",
          equipmentSlot: "Power",
        },
      },
    });
    expect(r.ok).toBe(false);
    if (!r.ok) expect(r.reasons[0]).toMatch(/Witherhoard|exotic/i);
  });

  it("allows legendary after one exotic weapon", () => {
    expect(
      assertSetExoticExclusivity({
        setType: "weapon",
        otherItems: [
          {
            slot: "primary",
            meta: { kind: "exotic_weapon", isExotic: true, name: "X" },
          },
        ],
        candidate: {
          slot: "special",
          meta: {
            kind: "weapon",
            isExotic: false,
            equipmentSlot: "Energy",
          },
        },
      }),
    ).toEqual({ ok: true });
  });

  it("allows replacing the slot that already holds the exotic", () => {
    // otherItems excludes the slot being replaced
    expect(
      assertSetExoticExclusivity({
        setType: "weapon",
        otherItems: [
          {
            slot: "special",
            meta: { kind: "weapon", isExotic: false },
          },
        ],
        candidate: {
          slot: "primary",
          meta: { kind: "exotic_weapon", isExotic: true },
        },
      }),
    ).toEqual({ ok: true });
  });

  it("blocks a second exotic armor piece", () => {
    const r = assertSetExoticExclusivity({
      setType: "armor",
      otherItems: [
        {
          slot: "helmet",
          meta: {
            kind: "exotic_armor",
            isExotic: true,
            name: "Synthoceps",
          },
        },
      ],
      candidate: {
        slot: "legs",
        meta: { kind: "exotic_armor", isExotic: true, name: "Dunemarchers" },
      },
    });
    expect(r.ok).toBe(false);
  });

  it("pair allows exotic weapon + exotic armor", () => {
    expect(
      assertSetExoticExclusivity({
        setType: "pair",
        otherItems: [
          {
            slot: "exotic_weapon",
            meta: { kind: "exotic_weapon", isExotic: true },
          },
        ],
        candidate: {
          slot: "exotic_armor",
          meta: { kind: "exotic_armor", isExotic: true },
        },
      }),
    ).toEqual({ ok: true });
  });

  it("assertSetCompositionAllowed combines fit + exclusivity", () => {
    const r = assertSetCompositionAllowed(
      "weapon",
      "heavy",
      {
        kind: "exotic_weapon",
        isExotic: true,
        equipmentSlot: "Power",
      },
      [
        {
          slot: "primary",
          meta: {
            kind: "exotic_weapon",
            isExotic: true,
            equipmentSlot: "Kinetic",
            name: "First",
          },
        },
      ],
    );
    expect(r.ok).toBe(false);
  });

  it("setAlreadyHasExotic detects weapon exotic", () => {
    const hit = setAlreadyHasExotic(
      "weapon",
      [
        {
          slot: "heavy",
          meta: { kind: "exotic_weapon", isExotic: true, name: "Ghorn" },
        },
      ],
      "weapon",
    );
    expect(hit?.meta.name).toBe("Ghorn");
  });
});

describe("helpers", () => {
  it("kindFromEntityStore maps stores", () => {
    expect(kindFromEntityStore("exotic-weapons")).toBe("exotic_weapon");
    expect(kindFromEntityStore("weapons")).toBe("weapon");
    expect(kindFromEntityStore("mods")).toBe("mod");
  });

  it("setItemMetaFromCatalog tags exotic weapons", () => {
    expect(
      setItemMetaFromCatalog({
        kind: "weapons",
        slot: "Power",
        isExotic: true,
      }),
    ).toEqual({
      kind: "exotic_weapon",
      equipmentSlot: "Power",
      isExotic: true,
    });
  });

  it("setItemMetaFromManifestCategory", () => {
    expect(setItemMetaFromManifestCategory("mods")).toMatchObject({ kind: "mod" });
    expect(setItemMetaFromManifestCategory("exotic-armor")).toMatchObject({
      kind: "exotic_armor",
      isExotic: true,
    });
  });
});

File: flutter/packages/domain/test/destiny_set_constraints_test.dart
import 'package:destiny2_domain/destiny2_domain.dart';
import 'package:test/test.dart';

/// Golden parity with TypeScript `destinySetConstraints.test.ts`.
void main() {
  group('assertSetItemAllowed', () {
    test('rejects invalid slots for set type', () {
      final r = assertSetItemAllowed(
        SetType.weapon,
        'helmet',
        const SetItemMeta(kind: SetItemKind.weapon, equipmentSlot: 'Kinetic'),
      );
      expect(r.ok, isFalse);
      expect(r.reasons.first, matches(RegExp('not valid', caseSensitive: false)));
    });

    test('rejects armor in a weapon set', () {
      final r = assertSetItemAllowed(
        SetType.weapon,
        'primary',
        const SetItemMeta(kind: SetItemKind.armor, equipmentSlot: 'Helmet'),
      );
      expect(r.ok, isFalse);
      expect(r.reasons.join(' '), matches(RegExp('weapons', caseSensitive: false)));
    });

    test('allows unknown kind on armor set (legendary armor store miss)', () {
      expect(
        assertSetItemAllowed(
          SetType.armor,
          'helmet',
          const SetItemMeta(kind: SetItemKind.unknown),
        ).ok,
        isTrue,
      );
    });

    test('rejects Kinetic weapon in special (Energy) slot', () {
      final r = assertSetItemAllowed(
        SetType.weapon,
        'special',
        const SetItemMeta(kind: SetItemKind.weapon, equipmentSlot: 'Kinetic'),
      );
      expect(r.ok, isFalse);
      expect(r.reasons.join(' '), matches(RegExp('Energy', caseSensitive: false)));
    });

    test('allows matching Kinetic weapon in primary', () {
      final r = assertSetItemAllowed(
        SetType.weapon,
        'primary',
        const SetItemMeta(
          kind: SetItemKind.weapon,
          equipmentSlot: 'Kinetic',
          isExotic: false,
        ),
      );
      expect(r.ok, isTrue);
    });

    test('allows exotic weapon in primary', () {
      final r = assertSetItemAllowed(
        SetType.weapon,
        'primary',
        const SetItemMeta(
          kind: SetItemKind.exoticWeapon,
          equipmentSlot: 'Kinetic',
          isExotic: true,
        ),
      );
      expect(r.ok, isTrue);
    });

    test('rejects wrong armor piece', () {
      final r = assertSetItemAllowed(
        SetType.armor,
        'helmet',
        const SetItemMeta(kind: SetItemKind.armor, equipmentSlot: 'Legs'),
      );
      expect(r.ok, isFalse);
    });

    test('allows matching helmet', () {
      expect(
        assertSetItemAllowed(
          SetType.armor,
          'helmet',
          const SetItemMeta(kind: SetItemKind.armor, equipmentSlot: 'Helmet'),
        ).ok,
        isTrue,
      );
    });

    test('rejects legendary weapon in pair exotic_weapon', () {
      final r = assertSetItemAllowed(
        SetType.pair,
        'exotic_weapon',
        const SetItemMeta(
          kind: SetItemKind.weapon,
          isExotic: false,
          equipmentSlot: 'Kinetic',
        ),
      );
      expect(r.ok, isFalse);
      expect(
        r.reasons.join(' '),
        matches(RegExp('exotic weapon', caseSensitive: false)),
      );
    });

    test('allows exotic weapon in pair exotic_weapon', () {
      expect(
        assertSetItemAllowed(
          SetType.pair,
          'exotic_weapon',
          const SetItemMeta(
            kind: SetItemKind.exoticWeapon,
            isExotic: true,
          ),
        ).ok,
        isTrue,
      );
    });

    test('rejects non-exotic armor in pair exotic_armor', () {
      final r = assertSetItemAllowed(
        SetType.pair,
        'exotic_armor',
        const SetItemMeta(
          kind: SetItemKind.armor,
          isExotic: false,
          equipmentSlot: 'Helmet',
        ),
      );
      expect(r.ok, isFalse);
    });

    test('rejects weapon hash meta in mod set', () {
      final r = assertSetItemAllowed(
        SetType.mod,
        'mod:1',
        const SetItemMeta(
          kind: SetItemKind.weapon,
          equipmentSlot: 'Kinetic',
        ),
      );
      expect(r.ok, isFalse);
    });

    test('allows mods in mod set', () {
      expect(
        assertSetItemAllowed(
          SetType.mod,
          'mod:42',
          const SetItemMeta(kind: SetItemKind.mod),
        ).ok,
        isTrue,
      );
    });

    test('allows fashion freely when slot valid', () {
      expect(
        assertSetItemAllowed(
          SetType.fashion,
          'ghost',
          const SetItemMeta(kind: SetItemKind.unknown),
        ).ok,
        isTrue,
      );
    });

    test('skips bucket check when equipmentSlot unknown', () {
      expect(
        assertSetItemAllowed(
          SetType.weapon,
          'primary',
          const SetItemMeta(kind: SetItemKind.weapon, equipmentSlot: null),
        ).ok,
        isTrue,
      );
    });
  });

  group('assertSetExoticExclusivity', () {
    test('blocks a second exotic weapon in a weapon set', () {
      final r = assertSetExoticExclusivity(
        setType: SetType.weapon,
        otherItems: const [
          SetOccupant(
            slot: 'primary',
            meta: SetItemMeta(
              kind: SetItemKind.exoticWeapon,
              isExotic: true,
              name: 'Witherhoard',
              equipmentSlot: 'Kinetic',
            ),
          ),
        ],
        candidate: const SetOccupant(
          slot: 'heavy',
          meta: SetItemMeta(
            kind: SetItemKind.exoticWeapon,
            isExotic: true,
            name: 'Gjallarhorn',
            equipmentSlot: 'Power',
          ),
        ),
      );
      expect(r.ok, isFalse);
      expect(
        r.reasons.first,
        matches(RegExp('Witherhoard|exotic', caseSensitive: false)),
      );
    });

    test('allows legendary after one exotic weapon', () {
      expect(
        assertSetExoticExclusivity(
          setType: SetType.weapon,
          otherItems: const [
            SetOccupant(
              slot: 'primary',
              meta: SetItemMeta(
                kind: SetItemKind.exoticWeapon,
                isExotic: true,
                name: 'X',
              ),
            ),
          ],
          candidate: const SetOccupant(
            slot: 'special',
            meta: SetItemMeta(
              kind: SetItemKind.weapon,
              isExotic: false,
              equipmentSlot: 'Energy',
            ),
          ),
        ).ok,
        isTrue,
      );
    });

    test('allows replacing the slot that already holds the exotic', () {
      // otherItems excludes the slot being replaced
      expect(
        assertSetExoticExclusivity(
          setType: SetType.weapon,
          otherItems: const [
            SetOccupant(
              slot: 'special',
              meta: SetItemMeta(kind: SetItemKind.weapon, isExotic: false),
            ),
          ],
          candidate: const SetOccupant(
            slot: 'primary',
            meta: SetItemMeta(
              kind: SetItemKind.exoticWeapon,
              isExotic: true,
            ),
          ),
        ).ok,
        isTrue,
      );
    });

    test('blocks a second exotic armor piece', () {
      final r = assertSetExoticExclusivity(
        setType: SetType.armor,
        otherItems: const [
          SetOccupant(
            slot: 'helmet',
            meta: SetItemMeta(
              kind: SetItemKind.exoticArmor,
              isExotic: true,
              name: 'Synthoceps',
            ),
          ),
        ],
        candidate: const SetOccupant(
          slot: 'legs',
          meta: SetItemMeta(
            kind: SetItemKind.exoticArmor,
            isExotic: true,
            name: 'Dunemarchers',
          ),
        ),
      );
      expect(r.ok, isFalse);
    });

    test('pair allows exotic weapon + exotic armor', () {
      expect(
        assertSetExoticExclusivity(
          setType: SetType.pair,
          otherItems: const [
            SetOccupant(
              slot: 'exotic_weapon',
              meta: SetItemMeta(
                kind: SetItemKind.exoticWeapon,
                isExotic: true,
              ),
            ),
          ],
          candidate: const SetOccupant(
            slot: 'exotic_armor',
            meta: SetItemMeta(
              kind: SetItemKind.exoticArmor,
              isExotic: true,
            ),
          ),
        ).ok,
        isTrue,
      );
    });

    test('assertSetCompositionAllowed combines fit + exclusivity', () {
      final r = assertSetCompositionAllowed(
        SetType.weapon,
        'heavy',
        const SetItemMeta(
          kind: SetItemKind.exoticWeapon,
          isExotic: true,
          equipmentSlot: 'Power',
        ),
        const [
          SetOccupant(
            slot: 'primary',
            meta: SetItemMeta(
              kind: SetItemKind.exoticWeapon,
              isExotic: true,
              equipmentSlot: 'Kinetic',
              name: 'First',
            ),
          ),
        ],
      );
      expect(r.ok, isFalse);
    });

    test('setAlreadyHasExotic detects weapon exotic', () {
      final hit = setAlreadyHasExotic(
        SetType.weapon,
        const [
          SetOccupant(
            slot: 'heavy',
            meta: SetItemMeta(
              kind: SetItemKind.exoticWeapon,
              isExotic: true,
              name: 'Ghorn',
            ),
          ),
        ],
        'weapon',
      );
      expect(hit?.meta.name, 'Ghorn');
    });
  });

  group('helpers', () {
    test('kindFromEntityStore maps stores', () {
      expect(kindFromEntityStore('exotic-weapons'), SetItemKind.exoticWeapon);
      expect(kindFromEntityStore('weapons'), SetItemKind.weapon);
      expect(kindFromEntityStore('mods'), SetItemKind.mod);
    });

    test('setItemMetaFromCatalog tags exotic weapons', () {
      final m = setItemMetaFromCatalog(
        kind: 'weapons',
        slot: 'Power',
        isExotic: true,
      );
      expect(m.kind, SetItemKind.exoticWeapon);
      expect(m.equipmentSlot, 'Power');
      expect(m.isExotic, isTrue);
    });

    test('setItemMetaFromManifestCategory', () {
      expect(
        setItemMetaFromManifestCategory('mods').kind,
        SetItemKind.mod,
      );
      final armor = setItemMetaFromManifestCategory('exotic-armor');
      expect(armor.kind, SetItemKind.exoticArmor);
      expect(armor.isExotic, isTrue);
    });

    test('shouldExcludeExoticFromSetCatalog hides second exotic', () {
      expect(
        shouldExcludeExoticFromSetCatalog(
          setType: SetType.weapon,
          targetSlot: 'heavy',
          otherItemsIncludingTarget: const [
            SetOccupant(
              slot: 'primary',
              meta: SetItemMeta(
                kind: SetItemKind.exoticWeapon,
                isExotic: true,
              ),
            ),
          ],
        ),
        isTrue,
      );
    });

    test('shouldExcludeExoticFromSetCatalog allows replace of exotic slot', () {
      expect(
        shouldExcludeExoticFromSetCatalog(
          setType: SetType.weapon,
          targetSlot: 'primary',
          otherItemsIncludingTarget: const [
            SetOccupant(
              slot: 'primary',
              meta: SetItemMeta(
                kind: SetItemKind.exoticWeapon,
                isExotic: true,
              ),
            ),
            SetOccupant(
              slot: 'special',
              meta: SetItemMeta(kind: SetItemKind.weapon, isExotic: false),
            ),
          ],
        ),
        isFalse,
      );
    });
  });
}
